lib/spack/spack/util/executable.py

```diff
@@ -269,39 +271,39 @@ def __str__(self):
         return " ".join(self.exe)
 
 
-@system_path_filter
 def which_string(*args, **kwargs):
     """Like ``which()``, but return a string instead of an ``Executable``."""
     path = kwargs.get("path", os.environ.get("PATH", ""))
     required = kwargs.get("required", False)
 
-    if isinstance(path, str):
-        path = path.split(os.pathsep)
+    if isinstance(path, list):
+        paths = [Path(str(x)) for x in path]
 
-    for name in args:
-        win_candidates = []
-        if sys.platform == "win32" and (not name.endswith(".exe") and not name.endswith(".bat")):
-            win_candidates = [name + ext for ext in [".exe", ".bat"]]
-        candidate_names = [name] if not win_candidates else win_candidates
+    if isinstance(path, str):
+        paths = [Path(x) for x in path.split(os.pathsep)]
 
+    for search_item in args:
+        is_cwd = True if search_item.startswith(".") else False
+        search_item = Path(search_item)
+        candidate_items = [Path(search_item)]
         if sys.platform == "win32":
-            new_path = path[:]
-            for p in path:
-                if os.path.basename(p) == "bin":
-                    new_path.append(os.path.dirname(p))
-            path = new_path
-
-        for candidate_name in candidate_names:
-            if os.path.sep in candidate_name:
-                exe = os.path.abspath(candidate_name)
-                if os.path.isfile(exe) and os.access(exe, os.X_OK):
-                    return exe
-            else:
-                for directory in path:
-                    directory = path_to_os_path(directory).pop()
-                    exe = os.path.join(directory, candidate_name)
-                    if os.path.isfile(exe) and os.access(exe, os.X_OK):
-                        return exe
+            if not search_item.suffix:
+                candidate_items = [
+                    search_item.parent / (search_item.name + ext) for ext in [".exe", ".bat"]
+                ]
+
+            paths_copy = paths
```

Comment:
This is kind of strange in the sense that it copies the reference, so the following iteration will iterate through the new items in `path` as they are created (i.e. there is no point to creating `paths_copy` as it exists here). I think it is clearer to say `paths_copy = list(paths)`

Reply:
That makes sense to do a full copy of the path. I refactored this to make it a bit clearer so this line doesn't completely apply anymore.